src/type.h

```diff
@@ -56,14 +57,20 @@ class Type {
   };
 
   Type() = default;  // Provided so Type can be member of a union.
-  Type(int32_t code) : enum_(static_cast<Enum>(code)) {}
-  Type(Enum e) : enum_(e) {}
+  Type(int32_t code) : enum_(static_cast<Enum>(code)), index_(UINT32_MAX) {}
+  Type(Enum e) : enum_(e), index_(UINT32_MAX)  {}
+  Type(Enum e, Index index) : enum_(e), index_(index) { assert(e == Enum::Reference); }
   operator Enum() const { return enum_; }
 
   bool IsRef() const {
     return enum_ == Type::ExternRef || enum_ == Type::FuncRef;
   }
 
+  bool IsFunctionReference() const {
```

Comment:
Its a little confusing to be me that `IsRef` would not return true for `Type::Reference` or that `IsFunctionReference` returns false for `Type::FuncRef`

Can you explain the different between the existing `Type::FuncRef` and this new `Type::Reference`.  Also aren't `Struct` and `Array` also a form of reference?    

Sorry I'm a little rusty on this stuff.

Reply:
Ah, I understand your confusion.
So, in typed function references proposal `funcref` and `externref` become `(ref null func)` and `(ref null extern)` respectively.

In WABT it will look like this `Type{Type::Reference, -0x10}` and `Type{Type::Reference, -0x11}`
but without TFR proposal it should look like `Type(Type::FuncRef)` and `Type(Type::ExternRef)`.
So, to support both variant I've updated `IsRef()` so that it returns `true` for both cases.

`IsReferenceWithIndex()` needed to distinguish between `FuncRef` and `ExternRef` because sadly
`(ref null func)` and `(ref null index)` have different encodings patterns and usage.